csharp/ql/lib/semmle/code/csharp/frameworks/Razor.qll

```diff
@@ -215,3 +215,61 @@ private class RelativeViewCallFilepath extends NormalizableFilepath {
   /** Holds if this string is the `idx`th path that will be searched for the `vc` call. */
   predicate hasViewCallWithIndex(ViewCall vc, int idx) { vc = vc_ and idx = idx_ }
 }
+
+/** A subclass of `Microsoft.AspNetCore.Mvc.RazorPages.PageModel` */
+class PageModelClass extends Class {
+  PageModelClass() {
+    this.getABaseType+().hasFullyQualifiedName("Microsoft.AspNetCore.Mvc.RazorPages", "PageModel")
+  }
+
+  /** Gets a handler method such as `OnGetAsync` */
+  Method getAHandlerMethod() {
+    result = this.getAMethod() and
+    result.getName().matches("On%") and
+    not exists(Attribute attr |
+      attr = result.getAnAttribute() and
+      attr.getType()
+          .hasFullyQualifiedName("Microsoft.AspNetCore.Mvc.RazorPages", "NonHandlerAttribute")
+    )
+  }
+
+  /** Gets the Razor Page that has this PageModel. */
+  RazorViewClass getPage() {
+    exists(Property modelProp |
+      modelProp.hasName("Model") and
+      modelProp.getType() = this and
+      modelProp.getDeclaringType() = result
+    )
+  }
+}
+
+private MethodCall getAPageCall(PageModelClass pm) {
+  result.getEnclosingCallable() = pm.getAHandlerMethod() and
+  result
+      .getTarget()
+      .hasFullyQualifiedName("Microsoft.AspNetCore.Mvc.RazorPages", "PageModel",
+        ["Page", "RedirectToPage"])
+}
+
+private MethodCall getImplicitThisCallInVoidHandler(PageModelClass pm) {
+  result.getEnclosingCallable() = pm.getAHandlerMethod() and
+  result.getEnclosingCallable().getReturnType() instanceof VoidType and
+  result.hasImplicitThisQualifier()
+}
+
+private class PageModelJumpNode extends DataFlow::NonLocalJumpNode {
+  PageModelClass pm;
+
+  PageModelJumpNode() {
+    this.asExpr() = [getAPageCall(pm), getImplicitThisCallInVoidHandler(pm)].getQualifier()
```

Comment:
For the ` getImplicitThisCallInVoidHandler` case, I would expect that the node should be the post-update node, as that is what is that target of stores. I.e., in `this.Name = source()` the store is into the post-update node for `this`.

Reply:
How can I target specifically the post-update node?

Reply:
You need to import `dataflow.internal.DataFlowPrivate.qll` and then do something like
```ql
private ThisAccess getThisCallInVoidHandler(PageModelClass pm) {
  exists(MethodCall mc |
    mc.getEnclosingCallable() = pm.getAHandlerMethod() and
    mc.getEnclosingCallable().getReturnType() instanceof VoidType and
    result = mc.getQualifier()
  )
}

private class PageModelJumpNode extends DataFlow::NonLocalJumpNode {
  PageModelClass pm;

  PageModelJumpNode() {
    this.asExpr() = getAPageCall(pm).getQualifier()
    or
    this.(PostUpdateNode).getPreUpdateNode().asExpr() = getThisCallInVoidHandler(pm)
  }
```

In case you are wondering, the test example works (by accident) because
```csharp
    public void OnGet() { 
        Name = source();
    }
```

is really

```csharp
    public void OnGet() { 
        this.Name = this.source();
    }
```

and there is use-use flow from the post-update `this` node in `this.Name` to the normal `this` node in `this.source()`, and then the current logic works.